include/drivers/pinctrl.h

```diff
@@ -0,0 +1,340 @@
+
+/** @endcond */
+
+/**
+ * @brief Define all pin control information for the given node identifier.
+ *
+ * This helper macro should be called together with device definition. It
+ * defines and initializes the pin control configuration for the device
+ * represented by node_id. Each pin control state (pinctrl-0, ..., pinctrl-N) is
+ * also defined and initialized.
+ *
+ * @param node_id Node identifier.
+ */
+#define PINCTRL_DT_DEFINE(node_id)					       \
+	Z_PINCTRL_STATES_DEFINE(node_id)				       \
+	Z_PINCTRL_CONFIG_ATTR struct pinctrl_config			       \
+	Z_PINCTRL_CONFIG_NAME(node_id) = Z_PINCTRL_CONFIG_INIT(node_id);
+
+/**
+ * @brief Define all pin control information for the given compatible index.
+ *
+ * @param inst Instance number.
+ *
+ * @see #PINCTRL_DT_DEFINE
+ */
+#define PINCTRL_DT_INST_DEFINE(inst) PINCTRL_DT_DEFINE(DT_DRV_INST(inst))
+
+/**
+ * @brief Obtain a reference to the pin control configuration given a node
+ * identifier.
+ *
+ * @param node_id Node identifier.
+ */
+#define PINCTRL_DT_CONFIG_GET(node_id) &Z_PINCTRL_CONFIG_NAME(node_id)
+
+/**
+ * @brief Obtain a reference to the pin control configuration given current
+ * compatible instance number.
+ *
+ * @param inst Instance number.
+ *
+ * @see #PINCTRL_DT_CONFIG_GET
+ */
+#define PINCTRL_DT_INST_CONFIG_GET(inst) \
+	PINCTRL_DT_CONFIG_GET(DT_DRV_INST(inst))
+
+/**
+ * @brief Find the state configuration for the given state id.
+ *
+ * @param config Pin controller configuration.
+ * @param id Pin controller state id.
+ * @param state Found state.
+ *
+ * @retval 0 If state has been found.
+ * @retval -ENOENT If the state has not been found.
+ */
+int pinctrl_state_lookup(const struct pinctrl_config *config,
+			 enum pinctrl_state id,
+			 const struct pinctrl_state_config **state);
+
+/**
+ * @brief Configure a set of pins
+ *
+ * @warning This function must never be used to configure pins used by an
+ * instantiated Zephyr driver.
+ *
+ * @param pins List of pins to be configured.
+ * @param pins_cnt Number of pins.
+ * @param reg Device register (optional, use #PINCTRL_REG_NONE if not used).
+ *
+ * @retval 0 If succeeded
+ * @retval -errno Negative errno for other failures.
+ */
+int pinctrl_configure_raw(const pinctrl_pin_t *pins,
+			  size_t pins_cnt, uintptr_t reg);
+
+/**
+ * @brief Set the pinctrl state for the given pinctrl device configuration.
+ *
+ * @warning Usage of this function is discouraged if runtime pin configuration
+ * has to be supported.
+ *
+ * @param config Pin control configuration.
+ * @param state State.
+ *
+ * @retval 0 If succeeded
+ * @retval -errno Negative errno for other failures.
+ */
+static inline int pinctrl_state_set_raw(const struct pinctrl_config *config,
+					const struct pinctrl_state_config *state)
+{
+	uintptr_t reg;
+
+#ifdef CONFIG_PINCTRL_STORE_REG
+	reg = config->reg;
+#else
+	reg = PINCTRL_REG_NONE;
+#endif
+
+	return pinctrl_configure_raw(state->pins, state->pins_cnt, reg);
```

Comment:
Can we pull iteration in this function instead of in `pinctrl_configure_raw`?

Reply:
From previous discussion:

> The reason is because in case of STM32 F1 or Nordic it is more efficient to do it all at once, in such cases a per-pin API would create unnecessary overhead
We can add it really, but then implementation should decide if the list version uses it or not
Actually, I think that for F1 would not even be feasible, since remap operation is per pin group really
ref. https://github.com/zephyrproject-rtos/zephyr/pull/37572/commits/4b7ee990ff2c16225eba6e9d1a9d01d72fe212bb#diff-f32e6b100432a089eed1035e0bc20b44906db32b69e9b3ab719bae2e19b88dcbR64